[BUG] `Py_UNICODE` is deprecated Python 3.13

### Describe the bug

While compiling using `3.13.0a3` getting error:

``` bash
Compiling src/liburing/file.pyx because it changed.
Compiling src/liburing/helper.pyx because it changed.                                                 
Compiling src/liburing/io_uring.pyx because it changed.    
Compiling src/liburing/liburing.pyx because it changed.
Compiling src/liburing/type.pyx because it changed.
Compiling src/liburing/version.pyx because it changed.                                               
[1/6] Cythonizing src/liburing/file.pyx
[2/6] Cythonizing src/liburing/helper.pyx                                                            
[3/6] Cythonizing src/liburing/io_uring.pyx                                                          
[4/6] Cythonizing src/liburing/liburing.pyx                                                          
[5/6] Cythonizing src/liburing/type.pyx                                                              
[6/6] Cythonizing src/liburing/version.pyx                                                           
running build_ext
building 'liburing.file' extension                                                                    
building 'liburing.helper' extension
creating build                            
creating build
creating build/temp.linux-x86_64-cpython-313
creating build/temp.linux-x86_64-cpython-313
building 'liburing.io_uring' extension
building 'liburing.liburing' extension
creating build/temp.linux-x86_64-cpython-313/src
building 'liburing.type' extension
creating build/temp.linux-x86_64-cpython-313/src
creating build/temp.linux-x86_64-cpython-313/src
building 'liburing.version' extension
creating build/temp.linux-x86_64-cpython-313/src/liburing
creating build/temp.linux-x86_64-cpython-313/src/liburing
creating build/temp.linux-x86_64-cpython-313/src/liburing
creating build/temp.linux-x86_64-cpython-313/src/liburing
creating build/temp.linux-x86_64-cpython-313/src/liburing
creating build/temp.linux-x86_64-cpython-313/src/liburing
gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall -fPIC -DCYTHON_TRACE_NOGIL=1 -Ilibs/liburing/src/include -I/opt/python/3.13/include/python3.13 -c src/liburing/io_uring.c -o build/temp.linux-x86_64-cpython-313/src/liburing/io_uring.o -Og
gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall -fPIC -DCYTHON_TRACE_NOGIL=1 -Ilibs/liburing/src/include -I/opt/python/3.13/include/python3.13 -c src/liburing/type.c -o build/temp.linux-x86_64-cpython-313/src/liburing/type.o -Og
gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall -fPIC -DCYTHON_TRACE_NOGIL=1 -Ilibs/liburing/src/include -I/opt/python/3.13/include/python3.13 -c src/liburing/helper.c -o build/temp.linux-x86_64-cpython-313/src/liburing/helper.o -Og
gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall -fPIC -DCYTHON_TRACE_NOGIL=1 -Ilibs/liburing/src/include -I/opt/python/3.13/include/python3.13 -c src/liburing/file.c -o build/temp.linux-x86_64-cpython-313/src/liburing/file.o -Og
gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall -fPIC -DCYTHON_TRACE_NOGIL=1 -Ilibs/liburing/src/include -I/opt/python/3.13/include/python3.13 -c src/liburing/liburing.c -o build/temp.linux-x86_64-cpython-313/src/liburing/liburing.o -Og
gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall -fPIC -DCYTHON_TRACE_NOGIL=1 -Ilibs/liburing/src/include -I/opt/python/3.13/include/python3.13 -c src/liburing/version.c -o build/temp.linux-x86_64-cpython-313/src/liburing/version.o -Og
src/liburing/helper.c:1308:1: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]         
 1308 | static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u)                      
      | ^~~~~~                                                                                        
src/liburing/helper.c: In function ‘__Pyx_Py_UNICODE_strlen’:                                         
src/liburing/helper.c:1310:5: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]         
 1310 |     const Py_UNICODE *u_end = u;                                                              
      |     ^~~~~                                                                                     
src/liburing/type.c:1322:1: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]           
 1322 | static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u)                      
      | ^~~~~~                                                                                        
src/liburing/type.c: In function ‘__Pyx_Py_UNICODE_strlen’:                                           
src/liburing/type.c:1324:5: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]           
 1324 |     const Py_UNICODE *u_end = u;                                                              
      |     ^~~~~                                                                                     
src/liburing/io_uring.c:1324:1: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]       
 1324 | static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u)                      
      | ^~~~~~                                                                                        
src/liburing/io_uring.c: In function ‘__Pyx_Py_UNICODE_strlen’:                                       
src/liburing/io_uring.c:1326:5: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]       
 1326 |     const Py_UNICODE *u_end = u;                                                              
      |     ^~~~~                                                                                     
src/liburing/version.c:1321:1: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]        
 1321 | static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u)                      
      | ^~~~~~                                                                                        
src/liburing/version.c: In function ‘__Pyx_Py_UNICODE_strlen’:                                        
src/liburing/version.c:1323:5: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]        
 1323 |     const Py_UNICODE *u_end = u;                                                              
      |     ^~~~~                                                                                     
src/liburing/file.c:1324:1: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]           
 1324 | static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u)                      
      | ^~~~~~                                                                                        
src/liburing/file.c: In function ‘__Pyx_Py_UNICODE_strlen’:                                           
src/liburing/file.c:1326:5: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]           
 1326 |     const Py_UNICODE *u_end = u;                                                              
      |     ^~~~~                                                                                     
src/liburing/liburing.c:1325:1: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]       
 1325 | static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u)                      
      | ^~~~~~                                                                                        
src/liburing/liburing.c: In function ‘__Pyx_Py_UNICODE_strlen’:                                       
src/liburing/liburing.c:1327:5: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]       
 1327 |     const Py_UNICODE *u_end = u;                                                              
      |     ^~~~~                                                                                     
creating build/lib.linux-x86_64-cpython-313
creating build/lib.linux-x86_64-cpython-313/liburing
gcc -shared build/temp.linux-x86_64-cpython-313/src/liburing/version.o -Llibs/liburing/src -luring -o build/lib.linux-x86_64-cpython-313/liburing/version.cpython-313-x86_64-linux-gnu.so
gcc -shared build/temp.linux-x86_64-cpython-313/src/liburing/helper.o -Llibs/liburing/src -luring -o build/lib.linux-x86_64-cpython-313/liburing/helper.cpython-313-x86_64-linux-gnu.so
gcc -shared build/temp.linux-x86_64-cpython-313/src/liburing/file.o -Llibs/liburing/src -luring -o build/lib.linux-x86_64-cpython-313/liburing/file.cpython-313-x86_64-linux-gnu.so
gcc -shared build/temp.linux-x86_64-cpython-313/src/liburing/liburing.o -Llibs/liburing/src -luring -o build/lib.linux-x86_64-cpython-313/liburing/liburing.cpython-313-x86_64-linux-gnu.so
gcc -shared build/temp.linux-x86_64-cpython-313/src/liburing/type.o -Llibs/liburing/src -luring -o build/lib.linux-x86_64-cpython-313/liburing/type.cpython-313-x86_64-linux-gnu.so
gcc -shared build/temp.linux-x86_64-cpython-313/src/liburing/io_uring.o -Llibs/liburing/src -luring -o build/lib.linux-x86_64-cpython-313/liburing/io_uring.cpython-313-x86_64-linux-gnu.so
copying build/lib.linux-x86_64-cpython-313/liburing/file.cpython-313-x86_64-linux-gnu.so -> src/liburing
copying build/lib.linux-x86_64-cpython-313/liburing/helper.cpython-313-x86_64-linux-gnu.so -> src/liburing
copying build/lib.linux-x86_64-cpython-313/liburing/io_uring.cpython-313-x86_64-linux-gnu.so -> src/liburing
copying build/lib.linux-x86_64-cpython-313/liburing/liburing.cpython-313-x86_64-linux-gnu.so -> src/liburing
copying build/lib.linux-x86_64-cpython-313/liburing/type.cpython-313-x86_64-linux-gnu.so -> src/liburing
copying build/lib.linux-x86_64-cpython-313/liburing/version.cpython-313-x86_64-linux-gnu.so -> src/liburing
```
Works fine for `3.8 - 3.12`

According to https://docs.python.org/3.13/c-api/unicode.html 
> "Note:
The [Py_UNICODE](https://docs.python.org/3.13/c-api/unicode.html#c.Py_UNICODE) representation has been removed since Python 3.12 with deprecated APIs. See [PEP 623](https://peps.python.org/pep-0623/) for more information."

### Code to reproduce the behaviour:

```cython
# example code
```


### Expected behaviour

_No response_

### OS

Linux

### Python version

3.13.0a3

### Cython version

3.0.8

### Additional context

_No response_

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -5500,7 +5500,7 @@ class SliceIndexNode(ExprNode):
                         start_code,
                         code.error_goto_if_null(result, self.pos)))
                 code.globalstate.use_utility_code(
-                          UtilityCode.load_cached("pyunicode_from_unicode", "StringTools.c"))
+                    UtilityCode.load_cached("pyunicode_from_unicode", "StringTools.c"))
             else:
                 code.putln(
                     "%s = __Pyx_PyUnicode_FromUnicodeAndLength(%s + %s, %s - %s); %s" % (
@@ -5511,7 +5511,7 @@ class SliceIndexNode(ExprNode):
                         start_code,
                         code.error_goto_if_null(result, self.pos)))
                 code.globalstate.use_utility_code(
-                          UtilityCode.load_cached("pyunicode_from_unicode", "StringTools.c"))
+                    UtilityCode.load_cached("pyunicode_from_unicode", "StringTools.c"))
 
         elif self.base.type is unicode_type:
             code.globalstate.use_utility_code(

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -5499,6 +5499,8 @@ class SliceIndexNode(ExprNode):
                         base_result,
                         start_code,
                         code.error_goto_if_null(result, self.pos)))
+                code.globalstate.use_utility_code(
+                          UtilityCode.load_cached("pyunicode_from_unicode", "StringTools.c"))
             else:
                 code.putln(
                     "%s = __Pyx_PyUnicode_FromUnicodeAndLength(%s + %s, %s - %s); %s" % (
@@ -5508,6 +5510,8 @@ class SliceIndexNode(ExprNode):
                         stop_code,
                         start_code,
                         code.error_goto_if_null(result, self.pos)))
+                code.globalstate.use_utility_code(
+                          UtilityCode.load_cached("pyunicode_from_unicode", "StringTools.c"))
 
         elif self.base.type is unicode_type:
             code.globalstate.use_utility_code(

--- a/Cython/Compiler/PyrexTypes.py
+++ b/Cython/Compiler/PyrexTypes.py
@@ -2685,7 +2685,7 @@ class CPointerBaseType(CType):
 
     def create_to_py_utility_code(self, env):
         if self.to_py_function == "__Pyx_PyUnicode_FromUnicode":
-            # Not included in the standard TypeConversions.c string conversions because Py_UNICODE 
+            # Not included in the standard TypeConversions.c string conversions because Py_UNICODE
             # is deprecated.
             env.use_utility_code(UtilityCode.load_cached("pyunicode_from_unicode", "StringTools.c"))
         return super().create_to_py_utility_code(env)

--- a/Cython/Compiler/PyrexTypes.py
+++ b/Cython/Compiler/PyrexTypes.py
@@ -1683,15 +1683,25 @@ class CType(PyrexType):
     #
 
     to_py_function = None
+    to_py_utility_code = None
     from_py_function = None
+    from_py_utility_code = None
     exception_value = None
     exception_check = 1
 
     def create_to_py_utility_code(self, env):
-        return self.to_py_function is not None
+        if self.to_py_function is not None:
+            if self.to_py_utility_code is not None:
+                env.use_utility_code(self.to_py_utility_code)
+            return True
+        return False
 
     def create_from_py_utility_code(self, env):
-        return self.from_py_function is not None
+        if self.from_py_function is not None:
+            if self.from_py_utility_code is not None:
+                env.use_utility_code(self.from_py_utility_code)
+            return True
+        return False
 
     def can_coerce_to_pyobject(self, env):
         return self.create_to_py_utility_code(env)
@@ -2679,16 +2689,11 @@ class CPointerBaseType(CType):
             self.exception_value = "NULL"
         elif self.is_pyunicode_ptr and not base_type.is_error:
             self.to_py_function = "__Pyx_PyUnicode_FromUnicode"
+            self.to_py_utility_code = UtilityCode.load_cached(
+                "pyunicode_from_unicode", "StringTools.c")
             if self.is_ptr:
                 self.from_py_function = "__Pyx_PyUnicode_AsUnicode"
             self.exception_value = "NULL"
-
-    def create_to_py_utility_code(self, env):
-        if self.to_py_function == "__Pyx_PyUnicode_FromUnicode":
-            # Not included in the standard TypeConversions.c string conversions because Py_UNICODE
-            # is deprecated.
-            env.use_utility_code(UtilityCode.load_cached("pyunicode_from_unicode", "StringTools.c"))
-        return super().create_to_py_utility_code(env)
 
     def py_type_name(self):
         if self.is_string:

--- a/Cython/Compiler/PyrexTypes.py
+++ b/Cython/Compiler/PyrexTypes.py
@@ -2683,6 +2683,13 @@ class CPointerBaseType(CType):
                 self.from_py_function = "__Pyx_PyUnicode_AsUnicode"
             self.exception_value = "NULL"
 
+    def create_to_py_utility_code(self, env):
+        if self.to_py_function == "__Pyx_PyUnicode_FromUnicode":
+            # Not included in the standard TypeConversions.c string conversions because Py_UNICODE 
+            # is deprecated.
+            env.use_utility_code(UtilityCode.load_cached("pyunicode_from_unicode", "StringTools.c"))
+        return super().create_to_py_utility_code(env)
+
     def py_type_name(self):
         if self.is_string:
             return "bytes"

--- a/Cython/Utility/StringTools.c
+++ b/Cython/Utility/StringTools.c
@@ -13,6 +13,7 @@
 static CYTHON_INLINE Py_ssize_t __Pyx_Py_UNICODE_ssize_strlen(const Py_UNICODE *u);/*proto*/
 
 //////////////////// ssize_pyunicode_strlen ////////////////////
+//@requires: pyunicode_strlen
 
 static CYTHON_INLINE Py_ssize_t __Pyx_Py_UNICODE_ssize_strlen(const Py_UNICODE *u) {
     size_t len = __Pyx_Py_UNICODE_strlen(u);
@@ -23,6 +24,27 @@ static CYTHON_INLINE Py_ssize_t __Pyx_Py_UNICODE_ssize_strlen(const Py_UNICODE *
     return (Py_ssize_t) len;
 }
 
+//////////////////// pyunicode_strlen.proto ///////////////
+
+// There used to be a Py_UNICODE_strlen() in CPython 3.x, but it is deprecated since Py3.3.
+static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u); /* proto */
+
+//////////////////// pyunicode_strlen /////////////////////
+
+// Note: will not work in the limited API since Py_UNICODE is not available there.
+// May stop working at some point after Python 3.13 (deprecated)
+static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u)
+{
+    const Py_UNICODE *u_end = u;
+    while (*u_end++) ;
+    return (size_t)(u_end - u - 1);
+}
+
+//////////////////// pyunicode_from_unicode.proto //////////////////////
+//@requires: pyunicode_strlen
+
+#define __Pyx_PyUnicode_FromUnicode(u)       PyUnicode_FromUnicode(u, __Pyx_Py_UNICODE_strlen(u))
+#define __Pyx_PyUnicode_FromUnicodeAndLength PyUnicode_FromUnicode
 
 //////////////////// InitStrings.proto ////////////////////
 
@@ -1161,3 +1183,4 @@ static CYTHON_INLINE PyObject* __Pyx_PyStr_Str(PyObject *obj) {
 
 #define __Pyx_PyObject_Str(obj) \
     (likely(PyString_CheckExact(obj)) ? __Pyx_NewRef(obj) : PyObject_Str(obj))
+

--- a/Cython/Utility/TypeConversion.c
+++ b/Cython/Utility/TypeConversion.c
@@ -1161,7 +1161,7 @@ static PyObject *__Pyx_c_func_ptr_to_capsule(__Pyx_generic_func_pointer funcptr,
     } else {
         // on all other platforms (which is the vast majority, since POSIX require a function pointer
         // can be  converted to a void*) we skip the allocation and store directly into the capsule's value.
-        // Use memcpy copy the data from the function pointer to the void*
+        // Use memcpy to copy the data from the function pointer to the void*.
         // (since just casting is prohibited by standard C, and using unions is prohibited by standard c++)
         void *copy_into;
         memcpy((void*)&copy_into, (void*)&funcptr, sizeof(funcptr));

--- a/Cython/Utility/TypeConversion.c
+++ b/Cython/Utility/TypeConversion.c
@@ -85,27 +85,7 @@ static CYTHON_INLINE PyObject* __Pyx_PyUnicode_FromString(const char*);
 #define __Pyx_PyByteArray_FromCString(s)   __Pyx_PyByteArray_FromString((const char*)s)
 #define __Pyx_PyStr_FromCString(s)     __Pyx_PyStr_FromString((const char*)s)
 #define __Pyx_PyUnicode_FromCString(s) __Pyx_PyUnicode_FromString((const char*)s)
-
-// There used to be a Py_UNICODE_strlen() in CPython 3.x, but it is deprecated since Py3.3.
-#if CYTHON_COMPILING_IN_LIMITED_API
-static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const wchar_t *u)
-{
-    const wchar_t *u_end = u;
-    while (*u_end++) ;
-    return (size_t)(u_end - u - 1);
-}
-#else
-static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u)
-{
-    const Py_UNICODE *u_end = u;
-    while (*u_end++) ;
-    return (size_t)(u_end - u - 1);
-}
-#endif
-
 #define __Pyx_PyUnicode_FromOrdinal(o)       PyUnicode_FromOrdinal((int)o)
-#define __Pyx_PyUnicode_FromUnicode(u)       PyUnicode_FromUnicode(u, __Pyx_Py_UNICODE_strlen(u))
-#define __Pyx_PyUnicode_FromUnicodeAndLength PyUnicode_FromUnicode
 #define __Pyx_PyUnicode_AsUnicode            PyUnicode_AsUnicode
 
 #define __Pyx_NewRef(obj) (Py_INCREF(obj), obj)
@@ -1181,7 +1161,7 @@ static PyObject *__Pyx_c_func_ptr_to_capsule(__Pyx_generic_func_pointer funcptr,
     } else {
         // on all other platforms (which is the vast majority, since POSIX require a function pointer
         // can be  converted to a void*) we skip the allocation and store directly into the capsule's value.
-        // Use memcpy copy the data from the function pointer to the void* 
+        // Use memcpy copy the data from the function pointer to the void*
         // (since just casting is prohibited by standard C, and using unions is prohibited by standard c++)
         void *copy_into;
         memcpy((void*)&copy_into, (void*)&funcptr, sizeof(funcptr));

